### DISCUSSION TITLE
Question about feat_static_cat

### DISCUSSION THREAD

**Discussion by NielsRogge** [awslabs/gluonts#1046] [Q&A] [answered]
In the [extended tutorial](https://gluon-ts.mxnet.io/examples/extended_forecasting_tutorial/extended_tutorial.html#2.1-Define-a-transformation), `feat_static_cat` is explained as: "static (over time) categorical features, list with dimension equal to the number of features"

What are examples of static categorical features? Are these features to identify individual time series, in case one wants to perform multivariate time series forecasting?

**Comment by timoschowski** [comment, marked as answer]
Example: If you have 5 different categorical features, the dimension needs to be 5.

Typically, you can have a categorical feature to identify groups of time series. E.g., all time series with categorical features A have a certain time of seasonality or a high volume or something, whereas time series with another feature behave differently.

**Comment by alexhallam** [comment]
Hi @timoschowski, I was just searching for `feat_static_cat` guidance also. If you don't mind I have a simple reproducible example that would clarify this for me. 

At the moment this trains fine, but it is not using `key` as a categorical feature. Could you show me how to modify this code to make that happen? 

```
from gluonts.dataset import common
from gluonts.model import deepar
from gluonts.model.deepar import DeepAREstimator
from gluonts.distribution.neg_binomial import NegativeBinomialOutput
from gluonts.trainer import Trainer

import pandas as pd
data = "https://gist.githubusercontent.com/alexhallam/f0410e546fde69e8bda92e89878386fe/raw/b582d71840de595e8329efec899ec3bfe17db32a/car_train.csv"
df = pd.read_csv(data)
df_cats = pd.get_dummies(df['key'], prefix='part')
data = common.ListDataset([{
    "start": df.date[0],
    "target": df.value,
    'feat_static_cat': df_cats
}],freq="d")
estimator = DeepAREstimator(
    prediction_length=6,
    freq="D",
    distr_output = NegativeBinomialOutput(),
    #use_feat_static_cat=True,
    trainer=Trainer(
        epochs=5
    )
)
predictor = estimator.train(training_data=data)
prediction = next(predictor.predict(data))
print(prediction.mean)
prediction.plot(output_file='graph.png')
```

**Comment by NielsRogge** [comment]
@alexhallam is this a single or multivariate time series?

**Comment by alexhallam** [comment]
Well, I know that I can model each of these series individually, but I also wanted to compare those results with using each time series key as a feature. A pooled vs non-pooled set up. What I am trying to do above is use each time series as a categorical feature.

**Comment by StatMixedML** [comment]
@alexhallam You might want to try the following for creating your dataset (using the code-snippet from above)

```python
from gluonts.dataset.common import ListDataset
from gluonts.dataset.field_names import FieldName

data_list = ListDataset(
    [
        {FieldName.START: pd.Timestamp(df_cats.iloc[:, col_idx].index[0], freq = "d"),
         FieldName.TARGET: df_cats.iloc[:, col_idx].values,
         FieldName.ITEM_ID: df_cats.columns[col_idx],
         FieldName.FEAT_STATIC_CAT: np.array([col_idx]),
        }
        for col_idx in range(df_cats.shape[1])
    ],
    freq="d"
)
```
It creates the dataset that can be used for model training. You can check out some of the entries of `gluonts.dataset.common.ListDataset` 

```python
data_list.list_data[0]
```
showing 

```python
{'start': Timestamp('1970-01-01 00:00:00', freq='D'),
 'target': array([0, 0, 0, ..., 0, 0, 0], dtype=uint8),
 'item_id': 'part_10055165',
 'feat_static_cat': array([0])}
```
The categorical feature `feat_static_cat`  corresponds to the numeric index of the loop, `item_id` is the column name. DeepAR learns an embedding for each categorical feature `feat_static_cat`.

**Reply by jonbeckman** [reply]
Hey @StatMixedML, I would appreciate if you could clarify something for me. In the example you provided, supplying `df_cats.shape[1]` or `n` number of time-series will result in `n` number of future time-series predicted (N:N)?

The reason I ask is because I am trying to understand the difference between your code snippet and something like this:
```
data_list = ListDataset(
    [
        {
            FieldName.START: pd.Timestamp(df_cats.iloc[:, 0].index[0], freq = "d"),
            FieldName.TARGET: df_cats.iloc[:, 0].values,
            FieldName.FEAT_DYNAMIC_REAL: [
                df[:, col_idx].values for col_idx in range(df_cats.shape[1]) if col_idx != 0
            ],
        }
    ],
    freq="d"
)
```
Where supplying `n` time-series will predict only one time-series (N:1)

**Comment by alexhallam** [comment]
This answers the question for me. What do you think @NielsRogge?

**Comment by ShirleyMgit** [comment]
Hi can use_feat_static_cat be used in deepVAR? it seems that the model does not recognize it

**Comment by mbohlkeschneider** [comment]
Hi @ShirleyMgit,

`DeepVAR` does not use `feat_static_cat` at the moment. It is a bit odd in the interface that you can still set it (and probably should be removed) but the network itself does not use it. In the multivariate case, it also only make sense if you have different multivariate time series that represent different categories. Most people simply use one, big multivariate dataset that includes all time series and there the concept of a categorical variable does not make sense (because all the time series are present at each forward pass). 

Hope that answers your question.

**Reply by ShirleyMgit** [reply]
Yes, in my problem I do have multiple multivariate timeseries that represent different categories. Is there any network that support it for multivariate timeseries? Is it possible for me to extand deepVAR to include it? Can I see the full source code? Thanks

**Comment by mbohlkeschneider** [comment]
Hi @ShirleyMgit, 

I see. I'm not aware of a model that does that in GluonTS. The DeepVAR model code is here: https://github.com/awslabs/gluonts/tree/dev/src/gluonts/mx/model/deepvar. If you add that, feel free to send us a PR!

**Reply by ShirleyMgit** [reply]
By the way, if there are no categorical variables why deepVAR gets 'cardinality' and 'embedding dimensions' as parameters?

**Reply by ShirleyMgit** [reply]
Hi,
I am trying to extend the network with an embedding layer for static features (similarly to deepAR).
Therefore, I am trying to understand the code.
Can you please explain what is the role of 'self.embed' and repeated_index_embeddings?
thanks

**Reply by ShirleyMgit** [reply]
also it seems that the code does not deal with the observed target in calculating the loss and it is not documented
